erts/emulator/nifs/common/zstd_nif.c

```diff
@@ -0,0 +1,1073 @@
+
+    if (ctx) {
+        reset_decompress_context(ctx);
+    }
+    return am_ok;
+}
+
+static ERL_NIF_TERM compress_close_nif(ErlNifEnv *env,
+                                      int argc,
+                                      const ERL_NIF_TERM argv[]) {
+    ZstdCtx *ctx = get_zstd_context(env, argv[0], compress_type);
+
+    if (ctx) {
+        close_compress_context(ctx);
+    }
+    return am_ok;
+}
+
+static ERL_NIF_TERM decompress_close_nif(ErlNifEnv *env,
+                                      int argc,
+                                      const ERL_NIF_TERM argv[]) {
+    ZstdCtx *ctx = get_zstd_context(env, argv[0], decompress_type);
+
+    if (ctx) {
+        close_decompress_context(ctx);
+    }
+    return am_ok;
+}
+
+static void compress_context_dtor(ErlNifEnv *env, void *ctx) {
+    close_compress_context(ctx);
+}
+
+static void decompress_context_dtor(ErlNifEnv *env, void *ctx) {
+    close_decompress_context(ctx);
+}
+
+static ERL_NIF_TERM create_cdict_nif(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
+    ErlNifBinary bin;
+    int level;
+
+    if (!enif_inspect_iolist_as_binary(env, argv[0], &bin) ||
+        !enif_get_int(env, argv[1], &level)) {
+        return enif_make_badarg(env);
+    } else {
+        ERL_NIF_TERM result;
+        ZstdDict *dict;
+        ZSTD_CDict *cdict = ZSTD_createCDict_advanced(bin.data, bin.size,
+                    ZSTD_dlm_byRef, ZSTD_dct_auto,
+                    ZSTD_getCParams(level, 0, bin.size),
+                    zstd_customMem);
+
+        if (!cdict)
+            return enif_make_tuple2(env, am_error,
+                enif_make_atom(env, "invalid_compress_dict"));
+
+        dict = enif_alloc_resource(compress_dict_type, sizeof(ZstdDict));
+
+        dict->c = cdict;
+        dict->env = enif_alloc_env();
+        enif_make_binary(dict->env, &bin);
+
+        result = enif_make_resource(env, (void *)dict);
+        enif_release_resource((void *)dict);
+        return result;
+    }
+}
+
+static void compress_dict_dtor(ErlNifEnv *env, void *ctx) {
+    ZstdDict *dict = ctx;
+    ZSTD_freeCDict(dict->c);
+    enif_free_env(dict->env);
+}
+
+static ERL_NIF_TERM create_ddict_nif(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
+    ErlNifBinary bin;
+
+    if (!enif_inspect_iolist_as_binary(env, argv[0], &bin)) {
+        return enif_make_badarg(env);
+    } else {
+        ERL_NIF_TERM result;
+        ZstdDict *dict;
+        ZSTD_DDict *ddict = ZSTD_createDDict_advanced(bin.data, bin.size,
+                    ZSTD_dlm_byRef, ZSTD_dct_auto, zstd_customMem);
+
+        if (!ddict)
+            return enif_make_tuple2(env, am_error,
+                enif_make_atom(env, "invalid_decompress_dict"));
+
+        dict = enif_alloc_resource(decompress_dict_type, sizeof(ZstdDict));
+
+        dict->d = ddict;
+        dict->env = enif_alloc_env();
+        enif_make_binary(dict->env, &bin);
```

Comment:
This will not work to keep `bin.data` alive if `argv[0]` is a non-contiguous iolist.
 `enif_inspect_iolist_as_binary` will allocate a TMP data buffer for `bin,` which is deallocated at return when `env` dies. Not a refc binary.

`enif_make_binary(dict->env, &bin)` will **copy** the data in `bin` as it's not a refc binary. Not keep the data in `bin` alive.

Maybe `enif_inspect_iolist_as_binary` should accept `env` as NULL and then call `enif_alloc_binary`  for `bin`. Similar as `enif_inspect_iovec`. Then we can release it with enif_release_binary and don't need the trick with `dict->env` and `enif_make_binary.`